src/backend/executor/nodeModifyTable.c

```diff
@@ -2506,3 +2448,79 @@ ExecSquelchModifyTable(ModifyTableState *node)
 			break;
 	}
 }
+
+/*
+ * Build a attribute mapping between child partition and the root partition in
+ * case if child partition has physically-different attribute numbers from
+ * root's due to dropped columns.
+ */
+void
+makePartitionCheckMap(EState *estate, ResultRelInfo *resultRelInfo)
+{
+	Relation	resultRelationDesc = resultRelInfo->ri_RelationDesc;
+	Oid			parentRelid;
+
+	Assert(PointerIsValid(estate->es_result_partitions));
+
+	/*
+	 * The partition selection operation is done with the parent partition's
+	 * attribute numbers, so if child partition has physically-different
+	 * attribute numbers due to dropped columns, we should map the child
+	 * attribute numbers to the parent's attribute numbers to perform the
+	 * partition selection. EState may not have the parent relation
+	 * information at the moment, so we have to do a hard job here by opening
+	 * it and compare the tuple descriptors. The purpose of this code is to
+	 * prevent the tuple from being incorrectly interpreted during partition
+	 * selection, that can be performed in ExecInsert, ExecDelete and
+	 * checkPartitionUpdate functions when we work with the leaf partition as
+	 * result relation.
+	 */
+	parentRelid = estate->es_result_partitions->part->parrelid;
+
+	if (resultRelInfo->ri_PartitionParent &&
+		parentRelid != RelationGetRelid(resultRelInfo->ri_PartitionParent))
+	{
+		resultRelInfo->ri_PartCheckTupDescMatch = 0;
+		if (resultRelInfo->ri_PartCheckMap != NULL)
+			pfree(resultRelInfo->ri_PartCheckMap);
+		if (resultRelInfo->ri_PartitionParent)
+			relation_close(resultRelInfo->ri_PartitionParent, AccessShareLock);
+	}
```

Comment:
in which case is this logic executed? even when working with a subpartition, parentRelid will refer to the root table.

Reply:
Probably it will never be executed. I'll remove that overhead code